graphbolt/include/graphbolt/csc_sampling_graph.h

```diff
@@ -13,9 +13,74 @@
 #include <string>
 #include <vector>
 
+// Add for RandomEngine.
+#include <dmlc/thread_local.h>
+
+#include <pcg_random.hpp>
+#include <random>
+#include <thread>
+
 namespace graphbolt {
 namespace sampling {
 
+namespace {
+
+// Get a unique integer ID representing this thread.
+inline uint32_t GetThreadId() {
+  static int num_threads = 0;
+  static std::mutex mutex;
+  static thread_local int id = -1;
+
+  if (id == -1) {
+    std::lock_guard<std::mutex> guard(mutex);
+    id = num_threads;
+    num_threads++;
+  }
+  return id;
+}
+
+};  // namespace
+
+/**
+ * @brief Thread-local Random Number Generator class
+ */
+class RandomEngine {
+ public:
+  /** @brief Constructor with default seed */
+  RandomEngine() {
+    std::random_device rd;
+    SetSeed(rd());
+  }
+
+  /** @brief Constructor with given seed */
+  explicit RandomEngine(uint64_t seed, uint64_t stream = GetThreadId()) {
+    SetSeed(seed, stream);
+  }
+
+  /** @brief Get the thread-local random number generator instance */
+  static RandomEngine* ThreadLocal() {
+    return dmlc::ThreadLocalStore<RandomEngine>::Get();
```

Comment:
Does it depend on current dgl?

Reply:
I don't think this implementation has anything to do with the current DGL, but only a few third-party libraries.